x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java

```diff
@@ -197,15 +213,29 @@ synchronized AsyncSearchResponse toAsyncSearchResponse(AsyncSearchTask task, lon
             searchResponse = null;
         } else {
             /*
+             * MP TODO: UPDATE THIS DOCUMENTATION TO MATCH NEW LOGIC
              * Build the response, reducing aggs if we haven't already and
              * storing the result of the reduction, so we won't have to reduce
              * the same aggregation results a second time if nothing has changed.
              * This does cost memory because we have a reference to the finally
              * reduced aggs sitting around which can't be GCed until we get an update.
              */
-            InternalAggregations reducedAggs = reducedAggsSource.get();
-            reducedAggsSource = () -> reducedAggs;
-            searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            if (searchResponseMerger == null) {
+                System.err.println("XXX DEBUG 1 - no search response merger - local aggs only");
+                InternalAggregations reducedAggs = reducedAggsSource.get();
+                reducedAggsSource = () -> reducedAggs;
+                searchResponse = buildResponse(task.getStartTimeNanos(), reducedAggs);
+            } else if (finalReduce == false) {
```

Comment:
Using this flag looks odd, mostly because this method should always ever expose finally reduced results, for user consumption.

Reply:
> this method should always ever expose finally reduced results, for user consumption.

No, that's not the case currently. Async search will show partially reduced results. For example, in the example below only 9 of 15 shards from the local cluster have reported back and we show those results (this example is from a test of code on `main` not this branch).

<details>
  <summary><b>Partial results of local cluster only</b> (toggle to view)</summary>

```
{
  "id": "FjNiUEszYnQ4U2thR1M1VjNOR0FVMEEaS2JNdEllR3hUVkdNMWh1aUozOF83UTo0MzA=",
  "is_partial": true,
  "is_running": true,
  "start_time_in_millis": 1702916562276,
  "expiration_time_in_millis": 1703348562276,
  "response": {
    "took": 11355,
    "timed_out": false,
    "terminated_early": false,
    "num_reduce_phases": 2,
    "_shards": {
      "total": 15,
      "successful": 9,  // 9/15 shards on the local cluster
      "skipped": 0,
      "failed": 0
    },
    "_clusters": {
      "total": 3,
      "successful": 0,
      "skipped": 0,
      "running": 3,
      "partial": 0,
      "failed": 0,
      "details": {
        "(local)": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        },
        "remote2": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        },
        "remote1": {
          "status": "running",
          "indices": "*",
          "timed_out": false
        }
      }
    },
    "hits": {
      "total": {
        "value": 752,
        "relation": "eq"
      },
      "max_score": null,
      "hits": []
    },
    "aggregations": {
      "indexgroup": {
        "doc_count_error_upper_bound": 0,
        "sum_other_doc_count": 0,
        "buckets": [
          {
            "key": "blogs",
            "doc_count": 737  // partial results - there are 1400 docs total
          },
          {
            "key": "web_traffic",
            "doc_count": 15
          }
        ]
      }
    }
  }
}

```
</details>

For CCS MRT=true, the final local cluster results are reported in two ways: via the onPartialReduce and onFinalReduce callbacks to the SearchProgressListener (coming into the `MutableSearchResponse.updatePartialResponse` method) and then the CCSActionListener for the local cluster sends the final full `SearchResponse` to the SearchResponseMerger.

So here we don't want to use the fully reduced results that came into `MutableSearchResponse.updatePartialResponse` because we are going to have them from the SearchResponseMerger list (otherwise those local results would be included twice in the merged response).

The logic here is:

```
if (searchResponseMerger == null) {
    // this is a local-only search or CCS MRT=false, so process the results
    // coming in from onPartialReduce as we do currently in 'main'
} else if (finalReduce == false) {
    // we are doing CCS MRT=true and the local cluster has reported back
    // only partial results (subset of shards), so do a SearchResponseMerger-
    // based merge of any full results that have come in from remote clusters 
    // and the partial results of the local cluster
} else {
    // we are doing CCS MRT=true and the local cluster has reported back
    // full results (via both (1) updatePartialResults and (2) the CCSActionListener
    // passing the full local-cluster SearchResponse to the SearchResponseMerger)
    // so create a merged result based on full responses held in SearchResponseMerger
    // received so far (may still be running, may be completed)
}
```

Reply:
> No, that's not the case currently. Async search will show partially reduced results.

Just to clarify what I meant: async search with show incremental results, but only finally reduced results are ever returned to the user via get async search.

Perhaps the confusion on my end is around naming. I understand now the difference between final reduction done locally, and updateFinalResponse which is called when all clusters are done, in the context of CCS. Perhaps we could rename the flag to localFinalReduce? One thing I am a bit fuzzy on though is whether we are ok ignoring partial aggs results if finalReduce is true. Doesn't the supplier get assigned anyways in updatePartialResponse ?

Reply:
> Perhaps we could rename the flag to localFinalReduce?
Good idea. Fixed in next push.

Reply:
> One thing I am a bit fuzzy on though is whether we are ok ignoring partial aggs results if finalReduce is true. 

I believe so. Once `localFinalReduce` is true (and CCS MRT=true), we no longer need to use the data held in `reducedAggsSource` because the final SearchResponse from the local cluster has been sent to the SearchResponseMerger (in TransportSearchAction, the local cluster [has its own CCSActionListener](https://github.com/elastic/elasticsearch/blob/0570b0baaa509436555c5711ba1e36d228e4da93/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L615)), so SearchResponseMerger.getMergedResponse will return the local results (hits and aggs). We no longer need to provide the local search response to it.